From 9f36d13258e82b2fdd97018239d8334761fef069 Mon Sep 17 00:00:00 2001
From: Mykola Lysenko <Mykola.Lysenko@amd.com>
Date: Thu, 12 May 2016 18:08:54 +0800
Subject: [PATCH 0482/1722] drm/amd/dal: fix memory issues in mst mgr

1) Fixes BSOD on MST device reconnect, by properly
manage dc sink ref counter;
2) Fixes memory leak of dc sink on driver unload.

Signed-off-by: Mykola Lysenko <Mykola.Lysenko@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c | 4 ++--
 drivers/gpu/drm/amd/dal/dc/core/dc.c                    | 2 ++
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c
index 7bbf6fc..e2857a5 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c
@@ -209,14 +209,13 @@ static struct dc_sink *dm_dp_mst_add_mst_sink(
 	if (edid_status != EDID_OK)
 		goto fail;
 
-	/* dc_sink_retain(&core_sink->public); */
-
 	return dc_sink;
 
 fail:
 	dc_link_remove_remote_sink(dc_link, dc_sink);
 
 fail_add_sink:
+        dc_sink_release(dc_sink);
 	return NULL;
 }
 
@@ -370,6 +369,7 @@ static void dm_dp_destroy_mst_connector(
 	aconnector->port = NULL;
 	if (aconnector->dc_sink) {
 		dc_link_remove_remote_sink(aconnector->dc_link, aconnector->dc_sink);
+		dc_sink_release(aconnector->dc_sink);
 		aconnector->dc_sink = NULL;
 	}
 	if (aconnector->edid) {
diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc.c b/drivers/gpu/drm/amd/dal/dc/core/dc.c
index 7d8d144..b9af925 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc.c
@@ -1288,6 +1288,8 @@ bool dc_link_add_remote_sink(const struct dc_link *link, struct dc_sink *sink)
 		return false;
 	}
 
+	dc_sink_retain(sink);
+
 	dc_link->remote_sinks[link->sink_count] = sink;
 	dc_link->sink_count++;
 
-- 
2.7.4

